vm/qemu/qemu.go

```diff
@@ -220,6 +224,56 @@ var linuxCmdline = []string{
 	"biosdevname=0",
 }
 
+func (inst *instance) qmp_recv(conn net.Conn) (map[string]json.RawMessage, error) {
+	var qmp map[string]json.RawMessage
+
+	dec := json.NewDecoder(conn)
+	err := dec.Decode(&qmp)
+
+	return qmp, err
+}
+
+func (inst *instance) qmp(cmd string) (map[string]json.RawMessage, error) {
+	addr := fmt.Sprintf("127.0.0.1:%v", inst.monport)
+	conn, err := net.Dial("tcp", addr)
+	if err != nil {
+		return nil, err
+	}
+
+	defer conn.Close()
+
+	if _, err = inst.qmp_recv(conn) ; err != nil {
+		return nil, err
+	}
+	conn.Write([]byte(`{"execute": "qmp_capabilities"}`))
+	if _, err = inst.qmp_recv(conn) ; err != nil {
+		return nil, err
+	}
+	conn.Write([]byte(cmd))
+	return inst.qmp_recv(conn)
+}
+
+func (inst *instance) hmp(cmd string, cpu int) (string, error) {
+	req := fmt.Sprintf(
+		`{"execute": "human-monitor-command",
+		"arguments": {"command-line": "%v",
+		"cpu-index": %v}}`, cmd, cpu)
+	resp, err := inst.qmp(req)
+	if err != nil {
+		return "", err
+	}
+
+	ret, ok := resp["return"]
+	if !ok {
+		if ret_err, ok := resp["error"]; ok {
+			return strconv.Unquote(string(ret_err))
+		}
+		return "", errors.New(fmt.Sprintf("No \"return\" nor \"error\" in [%v]", resp))
+	}
+
+	return strconv.Unquote(string(ret))
```

Comment:
This dealing with map and unquoting feels like we are under-using json package capabilities.
Can it be unmarshalled into?
```
struct {
  Error string
  Return json.RawMessage
}
```
What's in the Return?
strconv.Unquote will fail on arbitrary json strings, it's not supposed to work with json.
Is Return just a string as well? If so we can simply unmarshal into a struct with 2 strings.

Reply:
QMP returns json in "return" (and this is what qmp() returns as it does all this unmarshalling) but when HMP is wrapped into QMP, the return is always a string as HMP is user readable old-school format (H for "human").

Reply:
Then I think we should unmarshal directly into:
```
struct {
  Error string
  Return string
}
```

Reply:
```
type qmp_response_struct struct {
        Error string `json: error`
        Return string `json: return`
}
var qmp qmp_response_struct
dec := json.NewDecoder(conn)
err := dec.Decode(&qmp)
```
fails to decode
```
{"return": "VM status: running\r\n"}
```
with `decode = json: cannot unmarshal object into Go struct field qmp_response_struct.Return of type string { }`. I am really confused by this unmarshaller.

Reply:
it seems that `\r\n` confuses unmarshaller as `Return json.RawMessage` works.

Reply:
That is a standard-conforming json parser, I don't think it invents any own rules.

And it does unmarshal that data:
https://play.golang.org/p/J7-rfy2hj3- 

Btw, it also says the tags syntax is incorrect. You don't need tags here at all:
https://play.golang.org/p/lYvXo166GKa

Reply:
If it's unquoted string with literal new line and unterminated string on the line, then I guess it's invalid json.
I don't see use of json.RawMessage to help:
https://play.golang.org/p/8WdoElrZgcj

and just \n is equally wrong:
https://play.golang.org/p/ZPsYMnZxZEK

Reply:
nah, it is like this:  https://play.golang.org/p/KkdT0BnCoBl
```
package main

import (
	"encoding/json"
	"fmt"
	"strconv"
)

type qmp_response_struct struct {
	Return json.RawMessage
}

func main() {
	data := "{\"return\": \"VM status: running\\r\\n\"}"
	qmp := new(qmp_response_struct)
	err := json.Unmarshal([]byte(data), &qmp)
	s, err1 := strconv.Unquote(string(qmp.Return))
	fmt.Printf("qmp:=%v err=%v %v\n", s, err1, err)
}
```

Reply:
For this string it works without RawMessage and Unquote:
https://play.golang.org/p/EwHfsCCAQ5O
json itself handles escaping/unescaping when serializing to/from string. It should never needed to be done manually.

Reply:
Ah. I figured it out. The actual trafic is:
```
> {"execute": "qmp_capabilities"}
< {"QMP": {"version": {"qemu": {"micro": 50, "minor": 1, "major": 5}, "package": ""}, "capabilities": ["oob"]}}
{"return": {}}
> {"execute": "human-monitor-command", "arguments": {"command-line": "info status"}}
< {"return": "VM status: running\r\n"}
```
So I need to unmarshall "QMP" first to a struct and then I can start receiving what I expect. This way no need in  RawMessage (strangely it helped though but whatever).

Reply:
What about sending requests? Do you want me to use go's facilities to marshall structures to json or it can be strings?

Reply:
I don't have strong preference, but probably a bit nicer using json serialization:
https://play.golang.org/p/uNSQgU3NtYU